### DISCUSSION TITLE
Advanced usage example

### DISCUSSION THREAD

**Discussion by wholtz** [mamba-org/micromamba-docker#114] [Show and tell]
@maresb -- I wanted to show you a Dockerfile I made today that takes advantage of several of the enhancements you have contributed:

```Dockerfile
# this is a micromamba:0.19.1 image, using digest instead of tag for reproducibility
FROM mambaorg/micromamba@sha256:1ab2666fcd1559e9865d599433671f97cb735103af9351e5e10a3f0d5b6048c8

# Create 2 conda environments:
# "lab" environment will run the jupyterlab server
# "kernel" environment will be active in the "CondaKernel" kernel within a jupyter notebook

RUN micromamba create --yes --name lab --channel conda-forge \
       jupyterlab=3.2.6 \
       jupyterlab_widgets=1.0.2 \
       python=3.10.1 \
       xorg-libxrender=0.9.10 && \
    micromamba create --yes --name kernel --channel conda-forge \
       ipympl=0.8.5 \
       ipywidgets=7.6.5 \
       matplotlib=3.5.0 \
       python=3.10.1 && \
    micromamba clean --all --yes

# activate the "kernel" environment to be active within in the next RUN command
ARG MAMBA_DOCKERFILE_ACTIVATE=1 
ENV ENV_NAME=kernel

# create a kernel.json file for using the "kernel" conda environment in a notebook
RUN python -m ipykernel install --user --name conda-kernel --display-name CondaKernel

# activate the "lab" environment for when the CMD executes
ENV ENV_NAME=lab

EXPOSE 8888
WORKDIR /home/micromamba
CMD ["jupyter", "lab", "--ip=0.0.0.0", "--ServerApp.token=''", "--ServerApp.root_dir=/home/micromamba"]
```

This has been used for trouble shooting some interactions between JupyterLab server and the kernel environments used within a Jupyter notebook. I just wanted you to see your work being put to good use. Thanks!

**Comment by maresb** [comment]
That's really great, thanks so much for sharing! I'm glad to see it being put to good use.

**Comment by maresb** [comment]
Here is the sort of thing for which I've been using this image: [maresb/micromamba-docker](https://github.com/maresb/micromamba-devcontainer) plus the linked example.

It's currently a very rough draft. I think there are a few instances where I'm reimplementing some magic which VS Code already has, or doing things not-so-ideally. There's still a lot of work to be done, but I feel really good about using `mambaorg/micromamba` as the base image.

**Comment by wholtz** [comment]
Yikes! I see why you wanted a micromamba image with ubuntu as the base image. I'm quite close to having multiple base images ready to go.

I had not realized that `COPY --from ...` could source from an arbitrary image that is not a previous stage of the Dockerfile:
https://github.com/maresb/micromamba-devcontainer/blob/6c74b07740b13a229bba05d7368c5c1611e75d3b/Dockerfile#L16
It was useful for me to see that.

I found your `ENTRYPOINT` amusing:
https://github.com/maresb/micromamba-devcontainer/blob/6c74b07740b13a229bba05d7368c5c1611e75d3b/Dockerfile#L158
I wonder how many levels deep this pattern will eventually be taken by someone.....

**Reply by maresb** [reply]
Haha, I'm glad you're entertained by my hackery! And yes, that's _exactly_ why I asked about Ubuntu. It's very exciting that alternative base images are almost ready!

I've been maintaining in private a `jupyterlab/docker-stacks`-based dev container for quite a while. It became monolithic and very difficult to maintain. I had all sorts of hacks and workarounds. Fun!

This is my attempt to start from scratch using `micromamba-docker` + VS Code, sometimes incorporating ideas from my old hacks.

I should think about that `ENTRYPOINT`. At first I tried sourcing `_entrypoint.sh`, but that didn't work so well. :wink:

You might be interested in what I did with the metadata action in my [GH Actions workflow](https://github.com/maresb/micromamba-devcontainer/blob/main/.github/workflows/publish_image.yml)

I wanted to get something up-and-running last weekend, even if it was ugly. If I get it into better shape next weekend I may announce on mamba-org's Gitter to gauge interest. I'm interested in case you have any further thoughts or ideas.

**Reply by wholtz** [reply]
Yes, I am also making use of `docker/metadata-action`, but generating a few more tags than you are....

**Reply by maresb** [reply]
> Yes, I am also making use of `docker/metadata-action`

Where? I don't see it.

**Reply by maresb** [reply]
I had in mind that you could replace [this](https://github.com/mamba-org/micromamba-docker/blob/17cae43ef55329c42229e9760299fc6df4c7f031/.github/workflows/push_latest.yml#L16-L18) with [this](https://github.com/maresb/micromamba-devcontainer/blob/6c74b07740b13a229bba05d7368c5c1611e75d3b/.github/workflows/publish_image.yml#L35).

**Reply by wholtz** [reply]
On a local branch at the moment. I did some tagging tests over [here](https://github.com/wholtz/tag-test/blob/main/.github/workflows/build.yml).

**Reply by maresb** [reply]
The `tag-test` repo appears to be private. (I get 404'ed)

**Reply by maresb** [reply]
(I probably don't need to look in your testing repos though. I'm sure you understand what I mean.)

**Reply by wholtz** [reply]
oh right... tag-test is public now.

**Reply by maresb** [reply]
That's magnificent. I'm looking forward to seeing it in micromamba-docker.
